## GLOBAL FINTECH DIGITAL PAYMENTS DATASET PROJECT

## 1. Project Overview ##
In the digital payments industry, companies process millions of transactions daily across
multiple countries, currencies, payment channels, and customer segments. While the volume
of data generated is enormous, most businesses struggle to convert that data into decisions
that protect revenue, reduce risk, and improve customer experience.

This project simulates the work of a data analyst embedded in a fintech company's analytics
team. Using a realistic, messy dataset of **~100,000 payment transactions** spanning three
years across 12 countries, we will clean, analyse, and visualise the data to answer the
questions that actually matter to business stakeholders.


## 2. Dataset Overview
The Global FinTech Digital Payments Dataset is a synthetic but industry-realistic dataset modelled after real-world payment processor data,similar in structure to transaction logs from companies such as Stripe, PayPal, or African fintech providers like Flutterwave and Pesapal.

It is designed specifically for end-to-end data analytics practice, covering the full pipeline from raw messy data through cleaning, SQL analysis, exploratory data analysis, Power BI dashboarding, and stakeholder storytelling.


## 3. Dataset understanding
The dataset covers the digital payments and e-commerce financial transactions domain. The following industry contexts are represented:
•	Retail and e-commerce payment processing (online purchases across product categories)
•	Cross-border and multi-currency transactions with exchange rate data
•	Mobile and digital payment methods including mobile money, credit cards, and digital wallets
•	Fraud detection and risk management scenarios
•	Customer segmentation and lifetime value analysis


## STEPS
1. Loading the data to pandas
2. Explore and clean first
3. after cleaning save the cleaned data
4. connect to sql
5. write queries for the problems and business questions
6. visualize with matplotlib/seaborn
7. connect the csv to powerbi(or tableau) and do datamodeling, measures and dashboards
8. write a report(or ppt) with insights and recommendations

## 1. Step 1
- Loading data into pandas


In [2]:
import pandas as pd
import numpy as np
import sqlite3
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv('fintech_payments_raw.csv')
df


,transaction_id,transaction_date,customer_id,customer_age,customer_segment,country,payment_method,transaction_amount_usd,currency,exchange_rate,...,merchant_name,transaction_status,is_fraud,device_type,platform,session_duration_sec,chargeback_flag,discount_applied,net_revenue_usd,customer_lifetime_value
0,F40603CD-860A-4FC8,11-09-2022,CUST012749,30,Corporate,USA,PayPal,165.45,USD,1.00000,...,Berry-Chung,Completed,0,Mobile,Android,737.8,0,0.2507,123.97,792.52
1,3130BA47-B033-4DB1,2022/02/16,CUST003490,29,CORPORATE,France,Apple Pay,2417.58,EUR,1.08000,...,"Harrington, Case and Wagner",Complete,0,Tablet,Browser,80.9,0,0.2835,1732.20,97.8
2,38D4B31F-B69D-4F24,2022-12-26,CUST000423,34,SME,France,PAYPAL,34.81,EUR,1.08000,...,"Bennett, Thompson and Peters",NaN,0,Mobile,iOS,56.3,0,0.0272,33.86,493.36
3,FE4AF4FE-1436-4B2D,2024-05-26,CUST008754,29,Basic,Kenya,bank transfer,1.06,KES,0.00750,...,Gutierrez LLC,Declined,0,Mobile,Android,622.3,0,0.2939,0.75,423.85
4,6699804A-C105-4518,2023/11/06,CUST000018,32,SME,NG,PayPal,0.03,NGN,0.00130,...,Navarro and Sons,Completed,1,Desktop,iOS,1402.3,0,0.3575,0.02,969.81
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
101495,070C4FC3-C4B3-4FA0,2024-07-04,CUST011807,26,Corporate,Kenya,M-Pesa,6.76,KES,0.00750,...,"Kirk, Huffman and Eaton",success,0,Mobile,Android,77.6,0,0.1027,6.07,120.73
101496,54F66104-3706-4222,12-07-2023,CUST002306,35,Retail,ZA,Bank Transfer,2.53,ZAR,0.05400,...,Landry Inc,Completed,0,Desktop,USSD,61.7,0,0.1902,2.05,27775.3
101497,7EC43571-C117-42E0,2024-10-14,CUST007999,40,Retail,Nigeria,M-PESA,NaN,NGN,0.00130,...,"Meyers, Woods and Powell",Declined,0,Mobile,iOS,1047.9,0,0.0296,0.22,4834.32
101498,50A41DD8-F84A-4212,2022-11-23,CUST001786,21,Retail,Uganda,Credit Card,0.03,UGX,0.00027,...,Schultz Group,Completed,0,Desktop,Android,43.1,0,0.0486,0.03,115.69


## STEP 2
- Explore and clean

In [4]:
df = pd.read_csv('fintech_payments_raw.csv')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 101500 entries, 0 to 101499
Data columns (total 22 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   transaction_id           101500 non-null  str    
 1   transaction_date         101500 non-null  str    
 2   customer_id              101500 non-null  str    
 3   customer_age             101500 non-null  int64  
 4   customer_segment         101500 non-null  str    
 5   country                  101500 non-null  str    
 6   payment_method           101500 non-null  str    
 7   transaction_amount_usd   98455 non-null   float64
 8   currency                 101500 non-null  str    
 9   exchange_rate            97440 non-null   float64
 10  product_category         101500 non-null  str    
 11  merchant_id              101500 non-null  str    
 12  merchant_name            101500 non-null  str    
 13  transaction_status       98766 non-null   str    
 14  is_fraud       

In [5]:
df.head()

,transaction_id,transaction_date,customer_id,customer_age,customer_segment,country,payment_method,transaction_amount_usd,currency,exchange_rate,...,merchant_name,transaction_status,is_fraud,device_type,platform,session_duration_sec,chargeback_flag,discount_applied,net_revenue_usd,customer_lifetime_value
0,F40603CD-860A-4FC8,11-09-2022,CUST012749,30,Corporate,USA,PayPal,165.45,USD,1.0000,...,Berry-Chung,Completed,0,Mobile,Android,737.8,0,0.2507,123.97,792.52
1,3130BA47-B033-4DB1,2022/02/16,CUST003490,29,CORPORATE,France,Apple Pay,2417.58,EUR,1.0800,...,"Harrington, Case and Wagner",Complete,0,Tablet,Browser,80.9,0,0.2835,1732.20,97.8
2,38D4B31F-B69D-4F24,2022-12-26,CUST000423,34,SME,France,PAYPAL,34.81,EUR,1.0800,...,"Bennett, Thompson and Peters",NaN,0,Mobile,iOS,56.3,0,0.0272,33.86,493.36
3,FE4AF4FE-1436-4B2D,2024-05-26,CUST008754,29,Basic,Kenya,bank transfer,1.06,KES,0.0075,...,Gutierrez LLC,Declined,0,Mobile,Android,622.3,0,0.2939,0.75,423.85
4,6699804A-C105-4518,2023/11/06,CUST000018,32,SME,NG,PayPal,0.03,NGN,0.0013,...,Navarro and Sons,Completed,1,Desktop,iOS,1402.3,0,0.3575,0.02,969.81


- Columns with null values ; exhange rate, transaction_amount_usd,transaction_status, session_duration,discount_applied, customer_lifetime_value

In [6]:
df.describe()

,customer_age,transaction_amount_usd,exchange_rate,is_fraud,session_duration_sec,chargeback_flag,discount_applied,net_revenue_usd
count,101500.000000,98455.000000,97440.000000,101500.000000,99470.000000,101500.000000,99978.000000,101500.000000
mean,36.257852,920.387225,0.259562,0.022729,444.002447,0.008532,0.170976,69.944739
std,36.855236,16799.617711,0.453261,0.149039,690.466589,0.091974,0.188100,350.189738
min,-5.000000,0.000000,0.000270,0.000000,-499.960340,0.000000,0.000200,0.000000
25%,26.000000,0.280000,0.001300,0.000000,115.000000,0.000000,0.071500,0.230000
50%,35.000000,2.280000,0.012000,0.000000,241.000000,0.000000,0.134300,1.870000
75%,43.000000,23.100000,0.072000,0.000000,508.200000,0.000000,0.223475,18.750000
max,999.000000,499365.251012,1.270000,1.000000,21023.600000,1.000000,2.497202,28119.600000


In [7]:
df.tail()

,transaction_id,transaction_date,customer_id,customer_age,customer_segment,country,payment_method,transaction_amount_usd,currency,exchange_rate,...,merchant_name,transaction_status,is_fraud,device_type,platform,session_duration_sec,chargeback_flag,discount_applied,net_revenue_usd,customer_lifetime_value
101495,070C4FC3-C4B3-4FA0,2024-07-04,CUST011807,26,Corporate,Kenya,M-Pesa,6.76,KES,0.00750,...,"Kirk, Huffman and Eaton",success,0,Mobile,Android,77.6,0,0.1027,6.07,120.73
101496,54F66104-3706-4222,12-07-2023,CUST002306,35,Retail,ZA,Bank Transfer,2.53,ZAR,0.05400,...,Landry Inc,Completed,0,Desktop,USSD,61.7,0,0.1902,2.05,27775.3
101497,7EC43571-C117-42E0,2024-10-14,CUST007999,40,Retail,Nigeria,M-PESA,NaN,NGN,0.00130,...,"Meyers, Woods and Powell",Declined,0,Mobile,iOS,1047.9,0,0.0296,0.22,4834.32
101498,50A41DD8-F84A-4212,2022-11-23,CUST001786,21,Retail,Uganda,Credit Card,0.03,UGX,0.00027,...,Schultz Group,Completed,0,Desktop,Android,43.1,0,0.0486,0.03,115.69
101499,A4290350-C7E5-42CB,05/04/2024,CUST007353,44,Corporate,United States,Airtel Money,138.42,USD,1.00000,...,"Jensen, Arnold and Smith",Complete,0,Tablet,iOS,375.9,0,0.2548,103.15,363.78


## DATA CLEANING
- Data quality issues
- The following categories of messiness are embedded across the dataset:
•	Duplicates: Repeated transaction_id values (~1.5% of rows)
•	Missing values: Nulls in transaction_amount_usd, exchange_rate, transaction_status, and customer_lifetime_value
•	Outliers: Extreme values in customer_age (e.g. 999, -5), transaction_amount_usd, session_duration_sec, and discount_applied (>100%)
•	Inconsistent categories: Mixed casing and spelling variants in customer_segment, platform, payment_method, product_category, and country
•	Date format inconsistency: transaction_date stored in multiple formats across rows


In [8]:
#Dropping duplicates
df.duplicated().sum()
df = df.drop_duplicates()

In [9]:
df.duplicated().sum()

0

In [10]:
#Missing values/Null values
df.isnull().sum()


transaction_id                0
transaction_date              0
customer_id                   0
customer_age                  0
customer_segment              0
country                       0
payment_method                0
transaction_amount_usd     3045
currency                      0
exchange_rate              4058
product_category              0
merchant_id                   0
merchant_name                 0
transaction_status         2733
is_fraud                      0
device_type                   0
platform                      0
session_duration_sec       2030
chargeback_flag               0
discount_applied           1521
net_revenue_usd               0
customer_lifetime_value    5581
dtype: int64

In [11]:
#TRANSACTION AMOUNT COLUMN
df['transaction_amount_usd'].describe()

count     97565.000000
mean        928.036986
std       16875.830564
min           0.000000
25%           0.280000
50%           2.280000
75%          23.110000
max      499365.251012
Name: transaction_amount_usd, dtype: float64

In [12]:
df['transaction_amount_usd'] = df['transaction_amount_usd'].fillna(df['transaction_amount_usd'].median())

In [13]:
df['transaction_amount_usd'].isnull().sum()

0

In [14]:
df['session_duration_sec'].isnull().sum()

2030

In [15]:
##check for negative session durations
(df['session_duration_sec']<0).sum()

587

In [16]:
#fill the negative session durations with nan then with median
df.loc[df['session_duration_sec']<0,'session_duration_sec'] = np.nan
df['session_duration_sec'] = df['session_duration_sec'].fillna(df['session_duration_sec'].median())
(df['session_duration_sec']<0).sum()


0

In [17]:
#OUTLIERS IN THE AGE COLUMN
((df['customer_age']<18 )| (df['customer_age']>100)).sum()

500

In [18]:
#FILL THOSE COLUMNS WITH NAN THEN IMPUTE WITH MEDIAN AGE
df.loc[(df['customer_age']<18 )| (df['customer_age']>100),'customer_age'] = np.nan
df['customer_age'] = df['customer_age'].fillna(df['customer_age'].median())


In [19]:
df['customer_age'].isnull().sum()

0

In [20]:
#STANDARDIZING DATES
df['transaction_date'].head(10)

0     11-09-2022
1     2022/02/16
2     2022-12-26
3     2024-05-26
4     2023/11/06
5     2024-02-18
6    13-Sep-2022
7     2023-03-03
8     2023/12/10
9     27/11/2023
Name: transaction_date, dtype: str

In [21]:
# Convert using the flexible parser
df['transaction_date'] = pd.to_datetime(df['transaction_date'], format='mixed', errors='coerce')
df['transaction_date'].head(10)

0   2022-11-09
1   2022-02-16
2   2022-12-26
3   2024-05-26
4   2023-11-06
5   2024-02-18
6   2022-09-13
7   2023-03-03
8   2023-12-10
9   2023-11-27
Name: transaction_date, dtype: datetime64[us]

In [22]:
df['transaction_date'].isna().sum()

0

In [23]:
#TRANSACTION AMOUNT OUTLIERS AND NULLS
q1  = df['transaction_amount_usd'].quantile(0.25)
q3  = df['transaction_amount_usd'].quantile(0.75)
iqr = q3 - q1
upper_cap = q3 + 3 * iqr
iqr

df['transaction_amount_usd'] = df['transaction_amount_usd'].clip(upper_cap)
df['transaction_amount_usd'] = df['transaction_amount_usd'].fillna(df['transaction_amount_usd'].median())
df['transaction_amount_usd'].describe()

count    100610.000000
mean        964.681729
std       16615.790011
min          83.660000
25%          83.660000
50%          83.660000
75%          83.660000
max      499365.251012
Name: transaction_amount_usd, dtype: float64

In [24]:
#FIX DISCOUNT VALUES ABOVE 1 OR BELOW 0. (MEANING THEY ARE ABOVE 100% OR IN THE NEGATIVE)
extreme_disc = (df['discount_applied']>0).sum()
extreme_disc

99089

In [25]:
df['discount_applied'] = df['discount_applied'].clip(0, 1)
df['discount_applied'].fillna(df['discount_applied'].median())

0         0.2507
1         0.2835
2         0.0272
3         0.2939
4         0.3575
           ...  
101494    0.2746
101495    0.1027
101496    0.1902
101497    0.0296
101498    0.0486
Name: discount_applied, Length: 100610, dtype: float64

In [26]:
#MISSING EXCHANGE RATES (IMPUTING)
missing_rates = df['exchange_rate'].isnull().sum()
missing_rates

4058

In [27]:
df['exchange_rate'] = df.groupby('currency')['exchange_rate'].transform(lambda x : x.fillna(x.median()))


In [28]:
#STANDARDIZING FOUR COLUMNS IN ONE FUNCTION AND MAPPING (PAYMENT METHOD, CUSTOMER SEGMENT, COUNTRY,PLATFORM)
# 1. Automated Cleaning (Fixes casing, dashes, and extra spaces)
def auto_standardise(series):
    return (series
            .str.strip()
            .str.lower()
            .str.replace(r'[-_]', ' ', regex=True) # Turn 'm-pesa' into 'm pesa'
            .str.title()) # Turn 'm pesa' into 'M Pesa'

# 2. Minimalist Correction Map (Only for things logic can't fix)
special_cases = {
    'M Pesa': 'M-Pesa',
    'mpesa':'M-pesa',
    'Mpesa':'M-pesa',
    'Cc': 'Credit Card',
    'Dc': 'Debit Card',
    'Sme': 'SME',
    'Usa': 'United States',
    'Uk': 'United Kingdom'
}

# 3. Apply to the four columns
category_cols = ['payment_method', 'customer_segment', 'country', 'platform']

for col in category_cols:
    # First, run the automated logic
    df[col] = auto_standardise(df[col])
    # Then, only map the special cases
    df[col] = df[col].replace(special_cases)
    
    
    

In [29]:
df['payment_method'].head(10)

0           Paypal
1        Apple Pay
2           Paypal
3    Bank Transfer
4           Paypal
5        Apple Pay
6           M-Pesa
7       Debit Card
8           M-Pesa
9    Crypto Wallet
Name: payment_method, dtype: str

In [30]:
df.head(20)

,transaction_id,transaction_date,customer_id,customer_age,customer_segment,country,payment_method,transaction_amount_usd,currency,exchange_rate,...,merchant_name,transaction_status,is_fraud,device_type,platform,session_duration_sec,chargeback_flag,discount_applied,net_revenue_usd,customer_lifetime_value
0,F40603CD-860A-4FC8,2022-11-09,CUST012749,30.0,Corporate,United States,Paypal,165.450000,USD,1.0000,...,Berry-Chung,Completed,0,Mobile,Android,737.8,0,0.2507,123.97,792.52
1,3130BA47-B033-4DB1,2022-02-16,CUST003490,29.0,Corporate,France,Apple Pay,2417.580000,EUR,1.0800,...,"Harrington, Case and Wagner",Complete,0,Tablet,Browser,80.9,0,0.2835,1732.20,97.8
2,38D4B31F-B69D-4F24,2022-12-26,CUST000423,34.0,SME,France,Paypal,83.660000,EUR,1.0800,...,"Bennett, Thompson and Peters",NaN,0,Mobile,Ios,56.3,0,0.0272,33.86,493.36
3,FE4AF4FE-1436-4B2D,2024-05-26,CUST008754,29.0,Basic,Kenya,Bank Transfer,83.660000,KES,0.0075,...,Gutierrez LLC,Declined,0,Mobile,Android,622.3,0,0.2939,0.75,423.85
4,6699804A-C105-4518,2023-11-06,CUST000018,32.0,SME,Ng,Paypal,83.660000,NGN,0.0013,...,Navarro and Sons,Completed,1,Desktop,Ios,1402.3,0,0.3575,0.02,969.81
5,D8CBA630-CCE9-45E0,2024-02-18,CUST013384,49.0,Small Business,Nigeria,Apple Pay,83.660000,NGN,0.0013,...,"Franklin, Mills and Davis",Pending,0,Mobile,Point Of Sale,355.6,0,0.1889,1.67,2026.5
6,AD9392AE-AB5F-4B27,2022-09-13,CUST004139,34.0,Retail,Tanzania,M-Pesa,83.660000,TZS,0.0004,...,"Daugherty, Clark and Haynes",completed,0,Mobile,Web,361.9,0,0.1512,0.23,1938.3
7,9A369796-6C31-4644,2023-03-03,CUST011350,45.0,Corporate,Ghana,Debit Card,83.660000,GHS,0.0720,...,Paul-Nash,Complete,0,Desktop,Android,23.0,0,0.1285,1.59,NaN
8,B0FE7FAC-9B7A-4722,2023-12-10,CUST002354,37.0,SME,Ghana,M-Pesa,83.660000,GHS,0.0720,...,Mcclure and Sons,Completed,0,Desktop,Android,50.1,0,0.1365,5.55,129.13
9,2AC9649C-AE5B-492A,2023-11-27,CUST012667,27.0,SME,Nigeria,Crypto Wallet,83.660000,NGN,0.0013,...,Hogan-Cooper,Failed,0,Desktop,Android,51.7,0,0.0234,0.05,783.75


In [31]:
df['transaction_status'].value_counts()

transaction_status
Completed      58018
Failed          6280
Pending         5576
Declined        4784
SUCCESS         3759
success         3695
Complete        3668
completed       3554
Refunded        3316
Failure          507
failed           479
FAILED           476
PENDING          437
pending          431
In Progress      418
declined         369
DECLINED         355
Rejected         343
refunded         317
REFUNDED         312
?                123
check later      123
ok               117
error            111
not sure         106
TBD              105
done              98
Name: count, dtype: int64

In [32]:
#STANDARDIZE TRANSACTION STATUS
# 1. First, fix casing and spacing (Handles: SUCCESS, success, completed,  Pending )
df['transaction_status'] = auto_standardise(df['transaction_status'])

# 2. Map synonyms to canonical values
# This collapses different words that mean the same thing
status_synonyms = {
    'Success': 'Completed',
    'Complete': 'Completed',
    'Done': 'Completed',
    'Ok': 'Completed',
    'Failure': 'Failed',
    'Error': 'Failed',
    'In Progress': 'Pending',
    'Check Later': 'Pending',
    'Tbd': 'Pending',
    'Rejected': 'Declined'
}
df['transaction_status'] = df['transaction_status'].replace(status_synonyms)

# 3. Removing 'ok','tbd''etc' 
valid_statuses = ['Completed', 'Failed', 'Pending', 'Declined', 'Refunded']

df['transaction_status'] = df['transaction_status'].where(
    df['transaction_status'].isin(valid_statuses), 'Unknown'
)

# 4. Final safety fill for NaNs
df['transaction_status'] = df['transaction_status'].fillna('Unknown')

print(df['transaction_status'].value_counts())

transaction_status
Completed    72909
Failed        7853
Pending       7090
Declined      5851
Refunded      3945
Unknown       2962
Name: count, dtype: int64


In [33]:
# #INCONSISTENT CATEGORIES
#  #Consolidate into a single configuration
# mapping_config = {
#     'payment_method': payment_map,
#     'customer_segment': segment_map,
#     'platform': platform_map,
#     'product_category': category_map,
#     'country': country_map,
#     'transaction_status': status_map
# }

# #Using a loop to apply normalization to all columns
# for col, cmap in mapping_config.items():
#     if col in df.columns:
#         df[col] = df[col].str.strip().str.lower().map(cmap).fillna(df[col].str.strip())

In [34]:
#STANDARDIZE CUSTOMER LIFETIME VALUE COLUMN
# 1. Clean symbols and convert to numeric in one chain
df['customer_lifetime_value'] = (
    df['customer_lifetime_value']
    .replace(r'[\$,]', '', regex=True)  
    .pipe(pd.to_numeric, errors='coerce')
)

# 2. Impute missing values with the median
df['customer_lifetime_value'] = df['customer_lifetime_value'].fillna(df['customer_lifetime_value'].median())
df['customer_lifetime_value'].dtype


dtype('float64')

In [35]:
df.head(10)

,transaction_id,transaction_date,customer_id,customer_age,customer_segment,country,payment_method,transaction_amount_usd,currency,exchange_rate,...,merchant_name,transaction_status,is_fraud,device_type,platform,session_duration_sec,chargeback_flag,discount_applied,net_revenue_usd,customer_lifetime_value
0,F40603CD-860A-4FC8,2022-11-09,CUST012749,30.0,Corporate,United States,Paypal,165.45,USD,1.0000,...,Berry-Chung,Completed,0,Mobile,Android,737.8,0,0.2507,123.97,792.52
1,3130BA47-B033-4DB1,2022-02-16,CUST003490,29.0,Corporate,France,Apple Pay,2417.58,EUR,1.0800,...,"Harrington, Case and Wagner",Completed,0,Tablet,Browser,80.9,0,0.2835,1732.20,97.80
2,38D4B31F-B69D-4F24,2022-12-26,CUST000423,34.0,SME,France,Paypal,83.66,EUR,1.0800,...,"Bennett, Thompson and Peters",Unknown,0,Mobile,Ios,56.3,0,0.0272,33.86,493.36
3,FE4AF4FE-1436-4B2D,2024-05-26,CUST008754,29.0,Basic,Kenya,Bank Transfer,83.66,KES,0.0075,...,Gutierrez LLC,Declined,0,Mobile,Android,622.3,0,0.2939,0.75,423.85
4,6699804A-C105-4518,2023-11-06,CUST000018,32.0,SME,Ng,Paypal,83.66,NGN,0.0013,...,Navarro and Sons,Completed,1,Desktop,Ios,1402.3,0,0.3575,0.02,969.81
5,D8CBA630-CCE9-45E0,2024-02-18,CUST013384,49.0,Small Business,Nigeria,Apple Pay,83.66,NGN,0.0013,...,"Franklin, Mills and Davis",Pending,0,Mobile,Point Of Sale,355.6,0,0.1889,1.67,2026.50
6,AD9392AE-AB5F-4B27,2022-09-13,CUST004139,34.0,Retail,Tanzania,M-Pesa,83.66,TZS,0.0004,...,"Daugherty, Clark and Haynes",Completed,0,Mobile,Web,361.9,0,0.1512,0.23,1938.30
7,9A369796-6C31-4644,2023-03-03,CUST011350,45.0,Corporate,Ghana,Debit Card,83.66,GHS,0.0720,...,Paul-Nash,Completed,0,Desktop,Android,23.0,0,0.1285,1.59,667.48
8,B0FE7FAC-9B7A-4722,2023-12-10,CUST002354,37.0,SME,Ghana,M-Pesa,83.66,GHS,0.0720,...,Mcclure and Sons,Completed,0,Desktop,Android,50.1,0,0.1365,5.55,129.13
9,2AC9649C-AE5B-492A,2023-11-27,CUST012667,27.0,SME,Nigeria,Crypto Wallet,83.66,NGN,0.0013,...,Hogan-Cooper,Failed,0,Desktop,Android,51.7,0,0.0234,0.05,783.75


In [36]:
#RECALCULATE NET_REVENUE_USD AFTER CLEANING
df['net_revenue_usd'] = (df['transaction_amount_usd'] * (1 - df['discount_applied'])).round(2)
df['net_revenue_usd'].head()

0     123.97
1    1732.20
2      81.38
3      59.07
4      53.75
Name: net_revenue_usd, dtype: float64

In [37]:
df['discount_applied'] = df['discount_applied'].fillna(0)
df['discount_applied'].isnull().sum()


0

In [38]:
#RECALCULATE NET_REVENUE_USD AFTER CLEANING
df['net_revenue_usd'] = (df['transaction_amount_usd'] * (1 - df['discount_applied'])).round(2)
df['net_revenue_usd'].head()

0     123.97
1    1732.20
2      81.38
3      59.07
4      53.75
Name: net_revenue_usd, dtype: float64

In [39]:
remaining_nulls = df.isnull().sum()
remaining_nulls


transaction_id             0
transaction_date           0
customer_id                0
customer_age               0
customer_segment           0
country                    0
payment_method             0
transaction_amount_usd     0
currency                   0
exchange_rate              0
product_category           0
merchant_id                0
merchant_name              0
transaction_status         0
is_fraud                   0
device_type                0
platform                   0
session_duration_sec       0
chargeback_flag            0
discount_applied           0
net_revenue_usd            0
customer_lifetime_value    0
dtype: int64

## DONE WITH DATA CLEANING
- NEXT LOAD THE CLEANED DATA THEN CONNECT TO SQLITE3

In [40]:
# Save cleaned CSV
df.to_csv('fintech_payments_clean.csv', index=False)
print("Cleaned dataset saved → fintech_payments_clean.csv")

# Load into SQLite
conn = sqlite3.connect('fintech.db')
df.to_sql('transactions', conn, if_exists='replace', index=False)
print("Loaded into SQLite table 'transactions'")

Cleaned dataset saved → fintech_payments_clean.csv
Loaded into SQLite table 'transactions'


###  write queries for the problems and business questions

In [46]:
df_sql = pd.read_sql("""
SELECT name FROM sqlite_master WHERE type='table'
""", conn)

df_sql

,name
0,transactions


In [47]:
df=pd.read_sql("""
SELECT * FROM transactions
""",conn)
df.head(10)

,transaction_id,transaction_date,customer_id,customer_age,customer_segment,country,payment_method,transaction_amount_usd,currency,exchange_rate,...,merchant_name,transaction_status,is_fraud,device_type,platform,session_duration_sec,chargeback_flag,discount_applied,net_revenue_usd,customer_lifetime_value
0,F40603CD-860A-4FC8,2022-11-09 00:00:00,CUST012749,30.0,Corporate,United States,Paypal,165.45,USD,1.0000,...,Berry-Chung,Completed,0,Mobile,Android,737.8,0,0.2507,123.97,792.52
1,3130BA47-B033-4DB1,2022-02-16 00:00:00,CUST003490,29.0,Corporate,France,Apple Pay,2417.58,EUR,1.0800,...,"Harrington, Case and Wagner",Completed,0,Tablet,Browser,80.9,0,0.2835,1732.20,97.80
2,38D4B31F-B69D-4F24,2022-12-26 00:00:00,CUST000423,34.0,SME,France,Paypal,83.66,EUR,1.0800,...,"Bennett, Thompson and Peters",Unknown,0,Mobile,Ios,56.3,0,0.0272,81.38,493.36
3,FE4AF4FE-1436-4B2D,2024-05-26 00:00:00,CUST008754,29.0,Basic,Kenya,Bank Transfer,83.66,KES,0.0075,...,Gutierrez LLC,Declined,0,Mobile,Android,622.3,0,0.2939,59.07,423.85
4,6699804A-C105-4518,2023-11-06 00:00:00,CUST000018,32.0,SME,Ng,Paypal,83.66,NGN,0.0013,...,Navarro and Sons,Completed,1,Desktop,Ios,1402.3,0,0.3575,53.75,969.81
5,D8CBA630-CCE9-45E0,2024-02-18 00:00:00,CUST013384,49.0,Small Business,Nigeria,Apple Pay,83.66,NGN,0.0013,...,"Franklin, Mills and Davis",Pending,0,Mobile,Point Of Sale,355.6,0,0.1889,67.86,2026.50
6,AD9392AE-AB5F-4B27,2022-09-13 00:00:00,CUST004139,34.0,Retail,Tanzania,M-Pesa,83.66,TZS,0.0004,...,"Daugherty, Clark and Haynes",Completed,0,Mobile,Web,361.9,0,0.1512,71.01,1938.30
7,9A369796-6C31-4644,2023-03-03 00:00:00,CUST011350,45.0,Corporate,Ghana,Debit Card,83.66,GHS,0.0720,...,Paul-Nash,Completed,0,Desktop,Android,23.0,0,0.1285,72.91,667.48
8,B0FE7FAC-9B7A-4722,2023-12-10 00:00:00,CUST002354,37.0,SME,Ghana,M-Pesa,83.66,GHS,0.0720,...,Mcclure and Sons,Completed,0,Desktop,Android,50.1,0,0.1365,72.24,129.13
9,2AC9649C-AE5B-492A,2023-11-27 00:00:00,CUST012667,27.0,SME,Nigeria,Crypto Wallet,83.66,NGN,0.0013,...,Hogan-Cooper,Failed,0,Desktop,Android,51.7,0,0.0234,81.70,783.75


Countiries that generate the most revenue

In [54]:
df1=pd.read_sql("""
SELECT country, SUM(net_revenue_usd) AS Total_Revenue
                FROM transactions
                GROUP BY country
                ORDER BY Total_Revenue DESC
""",conn)
df1.head(10)

,country,Total_Revenue
0,Kenya,12972623.69
1,Nigeria,10509295.69
2,South Africa,9201904.38
3,United Kingdom,7616966.58
4,United States,6192515.67
5,Egypt,5022843.86
6,Uganda,4809573.09
7,Tanzania,4403731.44
8,Ghana,4268629.70
9,Germany,3870873.68


Countires that are underperforming

In [55]:
df1.tail(10)

,country,Total_Revenue
15,Eg,721039.29
16,Ug,704604.08
17,Tz,668260.61
18,De,503389.41
19,In,429030.32
20,Za,345555.55
21,Sa,309368.76
22,Us,200321.91
23,Gb,179019.45
24,Fr,112634.50
